src/rate_limiter.rs

```diff
@@ -136,16 +145,103 @@ impl RateLimiter {
+        let num_clients = if self.used.contains_key(client_ip) {
+            self.used.len()
+        } else {
+            self.used.len() + 1
+        };
+        cmp::max(MIN_CLIENT_CAPACITY, SOFT_CAPACITY / num_clients as u64)
+    }
+
+    /// Update a client's balance to compensate for initial over-counting.
+    ///
+    /// When a request is made, clients are charged at the maximum size of the data
+    /// being requested. This method compensates the client for the over-counting by
+    /// crediting them the difference between the maximum and the actual size of the response.
+    pub fn deduct_response(
+        &mut self,
+        client_ip: &IpAddr,
+        msg_id: &MessageId,
+        part_count: u32,
+        part_index: u32,
+        payload: &[u8],
+    ) -> Option<u64> {
+        use Response::*;
+
+        // Check that the response isn't a single-part response that we never overcharged for.
+        // This prevents a malicious client from discounting their GET by preceding
+        // it with a PUT with the same `msg_id`.
+        if part_count == 1 && part_index == 0 {
+            match serialisation::deserialise::<UserMessage>(payload) {
+                Ok(UserMessage::Response(response)) => {
+                    match response {
+                        // We overcharged for these, so we let them through.
+                        GetIData { .. } |
+                        GetAccountInfo { .. } |
+                        GetMData { .. } |
+                        GetMDataVersion { .. } |
+                        GetMDataShell { .. } |
+                        ListMDataEntries { .. } |
+                        ListMDataKeys { .. } |
+                        ListMDataValues { .. } |
+                        GetMDataValue { .. } |
+                        ListMDataPermissions { .. } |
+                        ListMDataUserPermissions { .. } |
+                        ListAuthKeysAndVersion { .. } => (),
+                        // These are responses to requests we didn't overcharge for.
+                        // All these responses *should* fit in a single part.
+                        PutIData { .. } |
+                        PutMData { .. } |
+                        MutateMDataEntries { .. } |
+                        SetMDataUserPermissions { .. } |
+                        DelMDataUserPermissions { .. } |
+                        ChangeMDataOwner { .. } |
+                        InsAuthKey { .. } |
+                        DelAuthKey { .. } => return None,
+                    }
+                }
+                _ => return None,
+            }
+        }
+
+        let amount_charged = match self.overcharged.remove(msg_id) {
+            Some(amount) => amount,
+            None => return None,
+        };
+
+        let deduction = amount_charged.saturating_sub(part_count as u64 * MAX_PART_LEN as u64);
+
+        self.used.get_mut(client_ip).map(|used| {
+            *used = used.saturating_sub(deduction);
```

Comment:
Well it seems it might not skew the logic after all as [client allowance](https://github.com/michaelsproul/routing/blob/944749dedc5a9c32cd958432d854e542e88be916/src/rate_limiter.rs#L150) is calculated after the call to update() which would have removed it (sort of a refresh) if necessary - but might be safer to call update after the refund too.

Anyway that brings to two corollaries:
1. Maybe we can simply check for the existence of client in the overcharged container before trying to execute the refund logic - that way we can save deserialising the payload unnecessarily.
1. there can be a potential disconnect between overcharge-container and used-container. If the responses never arrive, things continue to remain in the overcharge-container forever. So each time we [refrain from re-inserting into the used-container](https://github.com/michaelsproul/routing/blob/944749dedc5a9c32cd958432d854e542e88be916/src/rate_limiter.rs#L270), we should additionally get rid of such a client entry from overcharged-container too (so maybe store extra info there to tie it to the client info in used-container ?) ;; (Or make overcharged-container an LRU but not sure it's worth it)